### PR TITLE
API for configuration cache friendly execution

### PR DESCRIPTION
**Pull request by mlopatkin**
This PR provides configuration cache-friendly APIs to execute processes during the configuration phase. There are two options:
* The build author can provide a `ValueSource` implementation that receives `ExecOperations` instance and uses it to execute whatever is necessary. As for every `ValueSource`, its value becomes part of the configuration cache's fingerprint; the value is obtained for each build execution to check the validity of the cache. It is the build author's responsibility to ensure that the value is computed fast, computation doesn't invalidate the cache when not necessary and no unwanted side effects happen
* For simpler cases, two new methods are added to providers: `exec` and `javaexec`. Both allow specifying a command, similar to existing `Project.exec` and `Project.javaexec` tasks, but instead of running the command right away lazy providers of the command's exit code and outputs are returned. The providers can be used during configuration (so they become inputs of the configuration cache) or can be wired to task inputs as usual (so they become inputs of the execution phase). Again, the build author has to ensure that the command runs fast and produces no unwanted side effects

Other ways to run external processes during the configuration phase will eventually become configuration cache errors.

### CODE REVIEW

File: subprojects/model-core/src/main/java/org/gradle/api/internal/provider/sources/process/ProcessOutputValueSource.java [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+import javax.annotation.Nullable;
+import javax.inject.Inject;
+import java.io.ByteArrayOutputStream;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+public abstract class ProcessOutputValueSource implements ValueSource<ProcessOutputValueSource.ExecOutputData, ProcessOutputValueSource.Parameters>, Describable {
+    public interface Parameters extends ValueSourceParameters {
```
Comment (mlopatkin) [inline subprojects/model-core/src/main/java/org/gradle/api/internal/provider/sources/process/ProcessOutputValueSource.java:37]: I have to capture working dir there as well.

**Comment by mlopatkin** [comment]
@bot-gradle test this

**Comment by mlopatkin** [comment]
@bot-gradle test this

File: subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/fingerprint/ConfigurationCacheFingerprintWriter.kt [hunk 2]
```diff
@@ -339,6 +344,14 @@ class ConfigurationCacheFingerprintWriter(
         }
     }
 
+    private
+    fun reportExternalProcessOutputRead(executable: String) {
+        reportInput(consumer = null, documentationSection = null) {
```
Comment (eskatos) [inline subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/fingerprint/ConfigurationCacheFingerprintWriter.kt:349]: :x: This should link to the documentation section added in this very PR

File: subprojects/docs/src/snippets/valueProviders/externalProcessProvider/groovy/build.gradle [hunk 3]
```diff
@@ -0,0 +1,3 @@
+def gitVersion = providers.exec {
+    commandLine("git", "--version")
```
Comment (eskatos) [inline subprojects/docs/src/snippets/valueProviders/externalProcessProvider/groovy/build.gradle:2]: :lipstick: Maybe `git describe` would be a more actionable example? No strong opinion

**Review by eskatos — CHANGES_REQUESTED** [review]
LGTM overall :+1: 

Requested a single change wrt linking the doc in the input report.

One question: can I work with both stderr/stdout? for example to ask if any of the two contains a given string

File: subprojects/docs/src/snippets/valueProviders/externalProcessProvider/groovy/build.gradle (hunk 3 above)
Comment (mlopatkin) [reply subprojects/docs/src/snippets/valueProviders/externalProcessProvider/groovy/build.gradle:2]: I tried that but the test harness for samples has no git repo so the call just fails, failing the test.

Comment (eskatos) [reply subprojects/docs/src/snippets/valueProviders/externalProcessProvider/groovy/build.gradle:2]: I didn't think about that. The current samples are fine :+1:

**Comment by mlopatkin** [comment]
>One question: can I work with both stderr/stdout? for example to ask if any of the two contains a given string

So far you have to work on them separately, there is no merged stream. In theory, it can be useful for some cases (e.g. when the relative order of errors and outputs is meaningful), but it will increase the complexity of the resulting API, and buffers on the writing and reading sides can still mess things up anyway. So far you have to write a custom `ValueSource` to have this.

I was thinking about another approach with a pre-set output stream where you could do something like `setErrorOutput(getStandardOutput())` to tie streams together but I decided not to go this route, as it looks too confusing.

File: subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/fingerprint/ConfigurationCacheFingerprintWriter.kt (hunk 2 above)
Comment (mlopatkin) [reply subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/fingerprint/ConfigurationCacheFingerprintWriter.kt]: Nice catch, forgot about that. Fixed now.

**Comment by eskatos** [comment]
It would not work if you need to rely on the ordering but I assume you can `zip()` the two providers.

**Review by eskatos — APPROVED** [review]

File: subprojects/model-core/src/main/java/org/gradle/api/internal/provider/sources/process/DefaultExecOutput.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+public class DefaultExecOutput implements ExecOutput {
+    private final Provider<ExecOutputData> dataProvider;
+
+    public DefaultExecOutput(Provider<ExecOutputData> dataProvider) {
+        this.dataProvider = dataProvider;
+    }
+
+    @Override
+    public Provider<ExecResult> getResult() {
+        return dataProvider.map((Transformer<ExecResult, ExecOutputData> & Serializable) ExecOutputData::getResult);
```
Comment (bamboo) [inline subprojects/model-core/src/main/java/org/gradle/api/internal/provider/sources/process/DefaultExecOutput.java:37]: 💄 could use `SerializableLambdas.transformer` here, e.g., `dataProvider.map(transformer(ExecOutputData::getResult))`

Comment (mlopatkin) [reply subprojects/model-core/src/main/java/org/gradle/api/internal/provider/sources/process/DefaultExecOutput.java:37]: I was a bit reluctant to move `SerializableLambdas` out of `core` (which isn't accessible there) but I think having it in `base-services` alongside functional interfaces that it decorates should be fine in the end.

**Comment by mlopatkin** [comment]
@bot-gradle test this

**Comment by bamboo** [comment]
Just a heads-up about `TaskExecutionTracker`. I was in need of a similar service for the shared build services validation logic so I converted yours to Java and added the ability to track tasks, https://github.com/gradle/gradle/pull/19342/commits/c60312f9693486c4265f2f5a59ec2753a30abf43

I'll be happy to update the code in this PR once is merged to use the Java service.

**Comment by mlopatkin** [comment]
I'm going to land the whole stack in the [last PR](https://github.com/gradle/gradle/pull/19374), so I'm closing this one.

**Comment by mlopatkin** [comment]
> I'm going to land the whole stack in the [last PR](https://github.com/gradle/gradle/pull/19374), so I'm closing this one.

Didn't work, I cannot fix gradle build scripts without this and get self-tests to pass

**Comment by mlopatkin** [comment]
@bot-gradle test and merge

**Pull request merged**